### PR TITLE
Xcode Server Get Assets Action

### PR DESCRIPTION
**Pull request by czechboy0**
Added an action to download Integration assets from an Xcode Server instance running at a known host. This action downloads the assets into a specified folder and optionally cleans up everything except for the `.xcarchive`. This is super useful in cases when you use Xcode Server to run long-running tests (our build on multiple devices takes hours) and just want to automatically download the archive, sign and upload to the App Store.

I use this action from my dev Mac, but this could also run on the CI directly and be triggered by the post-build script, making the deployment automatic.

I do have an issue with a couple of tests which don't work because `Action.sh` doesn't actually do its thing in tests, so I don't know how to simulate the download, unzipping and copying of the `.tar.gz`. Let me know what you think I should do here @KrauseFx.

### CODE REVIEW

File: lib/fastlane/actions/xcode_server_get_assets.rb [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+      require 'excon'
+      require 'json'
+      require 'fileutils'
+
+      def self.run(params)
+
+        host = params[:host]
+        bot_name = params[:bot_name]
+        integration_number_override = params[:integration_number]
+        target_folder = params[:target_folder]
```
Comment (KrauseFx) [inline lib/fastlane/actions/xcode_server_get_assets.rb]: You generally don't need to copy those values into local variables: they are already duplicated and you can access/modify them without any performance loss :+1:

File: lib/fastlane/actions/xcode_server_get_assets.rb [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+        Excon.defaults[:ssl_verify_peer] = !trust_self_signed_certs # for self-signed certificates
+
+        # create Xcode Server config
+        xcs = XcodeServer.new(host, username, password)
+        bots = xcs.fetch_all_bots
+
+        Helper.log.info "Fetched #{bots.count} Bots from Xcode Server at #{host}.".yellow
+
+        # pull out names
+        bot_names = bots.map { |bot| bot['name'] }
```
Comment (KrauseFx) [inline lib/fastlane/actions/xcode_server_get_assets.rb]: This variable (`bot_names`) is unused

File: lib/fastlane/actions/xcode_server_get_assets.rb (hunk 1 above)
Comment (KrauseFx) [inline lib/fastlane/actions/xcode_server_get_assets.rb]: Of course, if you prefer it this way, I'm fine with it

File: lib/fastlane/actions/xcode_server_get_assets.rb [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+        found_bots = bots.select { |bot| bot['name'] == bot_name }
+        raise "Failed to find a Bot with name #{bot_name} on server #{host}".red if found_bots.count == 0
+
+        bot = found_bots[0]
+
+        Helper.log.info "Found Bot with name #{bot_name} with id #{bot['_id']}.".green
+
+        # we have our bot, get finished integrations, sorted from newest to oldest
+        integrations = xcs.fetch_integrations(bot['_id']).select { |i| i['currentStep'] == 'completed' }
+        raise "Failed to find any completed integration for Bot \"#{bot_name}\"".red if !integrations || integrations.count == 0
```
Comment (KrauseFx) [inline lib/fastlane/actions/xcode_server_get_assets.rb]: Change the condition to 
```ruby
... if (integrations || []).count == 0
```

File: lib/fastlane/actions/xcode_server_get_assets.rb [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+        
+        # if no integration number is specified, pick the newest one (this is sorted from newest to oldest)
+        if integration_number_override
+          integration = integrations.select { |i| i['number'] == integration_number_override }.first
+          raise "Specified integration number #{integration_number_override} does not exist.".red if !integration
```
Comment (KrauseFx) [inline lib/fastlane/actions/xcode_server_get_assets.rb]: Change `if !integration` to `unless integration`

File: lib/fastlane/actions/xcode_server_get_assets.rb [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+        else
+          integration = integrations.first
+        end
+
+        # consider: only taking the last successful one? or allow failing tests? warnings?
```
Comment (KrauseFx) [inline lib/fastlane/actions/xcode_server_get_assets.rb]: Ha, that's where you had the `TODO`? :stuck_out_tongue_winking_eye:

File: lib/fastlane/actions/xcode_server_get_assets.rb [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+
+        Helper.log.info "Using integration #{integration['number']}.".yellow
+
+        # fetch assets for this integration
+        assets_path = xcs.fetch_assets(integration['_id'], target_folder, self)
+        raise "Failed to fetch assets for integration #{ integration['number'] }." if !assets_path
```
Comment (KrauseFx) [inline lib/fastlane/actions/xcode_server_get_assets.rb]: Please remove the spaces inside `{}`

Comment (KrauseFx) [inline lib/fastlane/actions/xcode_server_get_assets.rb]: `unless` instead of `if !`

File: spec/fixtures/requests/xcs_folder/xcs_sample.log [hunk 7]
```diff
@@ -0,0 +1 @@
+I am a log
```
Comment (KrauseFx) [inline spec/fixtures/requests/xcs_folder/xcs_sample.log]: :+1:

File: lib/fastlane/actions/xcode_server_get_assets.rb [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+          raise "Failed to fetch Integrations for Bot #{bot_id} from Xcode Server at #{@host}, response: #{response.status}: #{response.body}".red if response.status != 200
+          integrations = JSON.parse(response.body)['results']
+        end
+
+        def fetch_assets(integration_id, target_folder, action)
+          
+          # create a temp folder and a file, stream the download into it
+          Dir.mktmpdir do |dir|
+
+            temp_file = File.join(dir, "tmp_download.#{rand(1000000)}")
```
Comment (KrauseFx) [inline lib/fastlane/actions/xcode_server_get_assets.rb]: You could also use `Time.now.to_i.to_s` if you assume the same action is not triggered multiple times a second.

File: lib/fastlane/actions/xcode_server_get_assets.rb [hunk 9]
```diff
@@ -0,0 +1,299 @@
...
+            action.sh "cd \"#{out_folder}\"; cat \"#{temp_file}\" | gzip -d | tar -x"
+
+            # then pull the real name from headers
+            asset_filename = response.headers['Content-Disposition'].split(';')[1].split('=')[1].gsub('"', '')
+            asset_foldername = asset_filename.split('.')[0]
+
+            # rename the folder in out_folder to asset_foldername
+            found_folder = Dir.entries(out_folder).select { |item| item != '.' && item != '..' }[0]
+
+            raise "Internal error, couldn't find unzipped folder".red if found_folder == nil
```
Comment (KrauseFx) [inline lib/fastlane/actions/xcode_server_get_assets.rb]: Change the condition to `found_folder.nil?`

File: lib/fastlane/actions/xcode_server_get_assets.rb [hunk 10]
```diff
@@ -0,0 +1,299 @@
...
+            out = File.join(target_folder, asset_foldername)
+            return out
+          end
+          return nil
+        end
+
+        def headers
+          require 'base64'
+          headers = { 
+            'User-Agent' => 'fastlane-xcode_server_get_assets', #XCS wants user agent. for some API calls. not for others. sigh.
```
Comment (KrauseFx) [inline lib/fastlane/actions/xcode_server_get_assets.rb]: Please add a space after `#`

File: lib/fastlane/actions/xcode_server_get_assets.rb [hunk 11]
```diff
@@ -0,0 +1,299 @@
...
+      #####################################################
+      # @!group Documentation
+      #####################################################
+
+      def self.description
+        "Downloads Xcode Bot assets like the `.xcarchive` and logs"
+      end
+
+      def self.details
+        "This action downloads assets from your Xcode Server Bot (works with Xcode Server
```
Comment (KrauseFx) [inline lib/fastlane/actions/xcode_server_get_assets.rb]: Did you try running `fastlane action xcode_server_get_assets`? Does that look okay? I think you might need `\n` in between.

**Comment by KrauseFx** [comment]
This looks great, thanks for working on it :+1: I added some notes. I can't test the integration myself, so I'm even more thankful for all the tests you added :rocket:

**Comment by czechboy0** [comment]
@KrauseFx No worries, I'll make those changes. What should I do about the tests needing `Action.sh` though?

File: lib/fastlane/actions/xcode_server_get_assets.rb (hunk 1 above)
Comment (czechboy0) [inline lib/fastlane/actions/xcode_server_get_assets.rb]: I like how once you define it, Sublime makes it super easy to use in multiple places below. But good to know, thanks! :blush:

File: lib/fastlane/actions/xcode_server_get_assets.rb (hunk 5 above)
Comment (czechboy0) [inline lib/fastlane/actions/xcode_server_get_assets.rb]: Yup. It's more a note for a possible future improvement.

File: lib/fastlane/actions/xcode_server_get_assets.rb (hunk 8 above)
Comment (czechboy0) [inline lib/fastlane/actions/xcode_server_get_assets.rb]: Yeah I guess, using `rand` feels good enough to me really though. Not sure what's more likely to clash.

File: lib/fastlane/actions/xcode_server_get_assets.rb (hunk 11 above)
Comment (czechboy0) [inline lib/fastlane/actions/xcode_server_get_assets.rb]: Yeah it looks alright I think https://www.dropbox.com/s/1ke7ho2weca8il8/Screenshot%202015-08-11%2016.20.23.png?dl=0

**Comment by czechboy0** [comment]
Ready for your :eyes: again.

**Comment by KrauseFx** [comment]
Looks great, thanks for working on this @czechboy0 :+1: :rocket:

**Comment by czechboy0** [comment]
Cool :+1:! Ping me when there's a release with this action in.

**Comment by KrauseFx** [comment]
I'll try to push a new version today :rocket:

**Pull request merged**